### PR TITLE
Add modern cryptors

### PR DESCRIPTION
**Pull request by olegbaturin**
| Q             | A
| ------------- | ---
| Is bugfix?    | ❌
| New feature?  | ✔️
| Breaks BC?    | ❌

### CODE REVIEW

File: src/PasswordHasher.php [hunk 1]
```diff
@@ -39,7 +39,7 @@ public function __construct(
         array|null $parameters = null,
     ) {
         if ($parameters === null) {
-            $this->parameters = self::SAFE_PARAMETERS[$this->algorithm] ?? [];
+            $this->parameters = self::SAFE_PARAMETERS[$this->algorithm ?? ''] ?? [];
```
Comment (vjik) [inline src/PasswordHasher.php:41]: @olegbaturin Please, extract this fix to separate PR.

File: src/Crypt/Cipher/OpenSSLAeadCipher.php [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+            throw new EncryptionException("Key must be {$this->keySize} bytes long.");
+        }
+        if (StringHelper::byteLength($nonce) !== $this->nonceSize) {
+            throw new EncryptionException("Nonce must be {$this->nonceSize} bytes long.");
+        }
+
+        $tag = StringHelper::byteSubstring($data, -self::TAG_SIZE);
+        $ciphertext = StringHelper::byteSubstring($data, 0, -self::TAG_SIZE);
+
+        $decrypted = openssl_decrypt($ciphertext, $this->cipher, $key, OPENSSL_RAW_DATA | OPENSSL_DONT_ZERO_PAD_KEY, $nonce, $tag);
```
Comment (samdark) [inline src/Crypt/Cipher/OpenSSLAeadCipher.php]: Truncated GCM tags issue. If `$data` is shorter than `TAG_SIZE`, we pass a 1-15 byte tag to `openssl_decrypt()`, and OpenSSL authenticates only the supplied tag bytes. That makes empty-plaintext forgery possible. One can search, for example, a 1-byte tag succeeds in about 256 attempts. Reject `$data` shorter than `self::TAG_SIZE` before slicing, and add tests for empty plaintext plus truncated tags.

File: composer.json [hunk 3]
```diff
@@ -37,6 +37,7 @@
         "php": "8.1 - 8.5",
         "ext-hash": "*",
         "ext-openssl": "*",
+        "ext-sodium": "*",
```
Comment (samdark) [inline composer.json]: Move to suggests.

File: src/Crypt/Cipher/SodiumAeadCipher.php [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+        'XChaCha20-Poly1305-IETF' => [SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_KEYBYTES, SODIUM_CRYPTO_AEAD_XCHACHA20POLY1305_IETF_NPUBBYTES],
+    ];
+
+    /**
+     * @param string $cipher The cipher to use (must be one of ALLOWED_CIPHERS keys).
+     *
+     * @throws RuntimeException If sodium extension is missing, cipher not allowed, or AES-256-GCM without hardware support.
+     */
+    public function __construct(
+        private readonly string $cipher = 'AES-256-GCM',
```
Comment (samdark) [inline src/Crypt/Cipher/SodiumAeadCipher.php]: Is it a good default since there could be no hardware support? In the README, the cipher is not configured, so the following examples will result in an exception.

How about choosing default such as `XChaCha20-Poly1305-IETF`?

File: README.md [hunk 5]
```diff
@@ -167,6 +128,213 @@ There is a special function in PHP that compares strings in a constant time:
...
+],
+```
+
+#### KdfPassword - for low‑entropy passwords
+This first applies `PBKDF2` with a configurable iteration count, then `HKDF` to derive the final key.
+Follow OWASP recommendations for iteration counts.
+
+```php
+// /config/di.php
+use Yiisoft\Security\Crypt\Kdf\KdfPassword;
```
Comment (samdark) [inline README.md]: KdfPasswordPbkdf2 / KdfPasswordArgon2?

File: README.md [hunk 6]
```diff
@@ -167,6 +128,213 @@ There is a special function in PHP that compares strings in a constant time:
...
+
+KdfPassword::class => [
+    '__construct()' => [
+        'algorithm' => 'sha512', // any hash_hmac_algos()
```
Comment (samdark) [inline README.md]: `hashAlgo`?

File: src/Crypto/VersionedCryptor.php [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+
+    public function encrypt(
+        string $data,
+        #[SensitiveParameter]
+        string $secret,
+        string $context = ''
+    ): string {
+        $payload = $this->cryptors[$this->currentVersion]->encrypt($data, $secret, $context);
+
+        return $this->currentVersion . $payload;
```
Comment (samdark) [inline src/Crypto/VersionedCryptor.php:72]: Metadata is not being authenticated. Shouldn't it?

File: src/Crypt/EnvelopeCryptor.php [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+        $dek = random_bytes($this->keySize);
+        $dekNonce = random_bytes($this->nonceSize);
+        $dataNonce = random_bytes($this->nonceSize);
+
+        $kek = $this->kdf->createKey($secret, $this->keySize, $context, $keySalt);
+        $dekEncrypted = $this->cipher->encrypt($dek, $kek, $dekNonce);
+        $dataEncrypted = $this->cipher->encrypt($data, $dek, $dataNonce);
+
+        // keySalt || dekNonce || cipherdek || tag || dataNonce || ciphertext || tag
+        return $keySalt . $dekNonce . $dekEncrypted . $dataNonce . $dataEncrypted;
```
Comment (samdark) [inline src/Crypt/EnvelopeCryptor.php]: Metadata is not authenticated. Shouldn't it?

**Review by samdark — CHANGES_REQUESTED** [review]
Not enough negative/security tests around metadata tampering: version byte changes, salt/nonce field mutation, truncated payloads at every boundary, truncated tags, empty plaintext, wrong context, and cross-version/cross-cryptor confusion.

File: src/Crypt/Kdf/KdfPasswordArgon2.php [hunk 9]
```diff
@@ -0,0 +1,90 @@
...
+     */
+    public function createKey(
+        #[SensitiveParameter]
+        string $secret,
+        int $keySize,
+        string $context,
+        string $salt,
+    ): string {
+        try {
+            $key = sodium_crypto_pwhash($keySize, $secret, $salt, $this->opslimit, $this->memlimit, $this->algo);
```
Comment (samdark) [inline src/Crypt/Kdf/KdfPasswordArgon2.php]:
```suggestion
            $key = sodium_crypto_pwhash(32, $secret, $salt, $this->opslimit, $this->memlimit, $this->algo);
```

This size should not depend on algo.

File: src/Crypt/Kdf/KdfPasswordPbkdf2.php [hunk 10]
```diff
@@ -0,0 +1,85 @@
...
+     */
+    public function createKey(
+        #[SensitiveParameter]
+        string $secret,
+        int $keySize,
+        string $context,
+        string $salt,
+    ): string {
+        try {
+            $key = hash_pbkdf2($this->hashAlgo, $secret, $salt, $this->iterations, $keySize, true);
```
Comment (samdark) [inline src/Crypt/Kdf/KdfPasswordPbkdf2.php]:
```suggestion
            $key = hash_pbkdf2($this->hashAlgo, $secret, $salt, $this->iterations, 32, true);
```

This size should not depend on algo.

**Review by samdark — CHANGES_REQUESTED** [review]

File: src/Crypto/VersionedCryptor.php (hunk 7 above)
Comment (olegbaturin) [reply src/Crypto/VersionedCryptor.php:72]: No, authentication is not needed, because that is the task of the cryptor being used. The task of VersionedCryptor is to add a version string.

File: src/Crypt/EnvelopeCryptor.php (hunk 8 above)
Comment (olegbaturin) [reply src/Crypt/EnvelopeCryptor.php]: Authentication should be performed directly within the encryption algorithm, meaning an algorithm with AEAD should be used. I have added this to the description of all cryptors. If encryption without AEAD is used, then the authentication tag can be handled either by the cipher or the cryptor, but that would be a separate implementation.

File: src/Crypt/Cipher/SodiumAeadCipher.php (hunk 4 above)
Comment (olegbaturin) [reply src/Crypt/Cipher/SodiumAeadCipher.php]: In general, yes, `XChaCha20-Poly1305-IETF` is recommended by default because of the 24-byte nonce size. However, considering its use in key/password derivation algorithms, it may be excessive; one could use `ChaCha20-Poly1305-IETF`.

In that case, perhaps we should set `ChaCha20-Poly1305-IETF` as the default in both the `OpenSSLAeadCipher` and `SodiumAeadCipher` classes?

File: composer.json (hunk 3 above)
Comment (olegbaturin) [reply composer.json]: both openssl and sodium extensions moved to suggests

**Comment by olegbaturin** [comment]
> Not enough negative/security tests around metadata tampering: version byte changes, salt/nonce field mutation, truncated payloads at every boundary, truncated tags, empty plaintext, wrong context, and cross-version/cross-cryptor confusion.

>version byte changes

`VersionedCryptorTest::testDecryptThrowsExceptionWhenVersionNotFound()`

>salt field mutation

`AbstractKdfCase::testDifferentParamsProducesDifferentKey()`
+
`AbstractCipherCase::testDecryptWithWrongKeyThrowsException()`

>nonce field mutation

`CipherWithNounceTrait::testDecryptWithWrongNonceThrowsException()`

>truncated payloads at every boundary

wrong/tempered data tests for kdf/cipher

>truncated tags

`CipherWithAeadTrait`

>empty plaintext

Added tests for ciphers.
Makes no sense for cryptors, because AEAD ciphers always return ciphertext with an embedded tag.

>wrong context

`AbstractKdfCase::testDifferentParamsProducesDifferentKey()`
+
`AbstractCipherCase::testDecryptWithWrongKeyThrowsException()`
 
>cross-version confusion

`VersionedCryptorTest::testDecryptThrowsExceptionWhenVersionNotFound()`

>cross-cryptor confusion

wrong/tempered data tests for kdf/cipner

File: src/Crypt/Cipher/OpenSSLAeadCipher.php (hunk 2 above)
Comment (olegbaturin) [reply src/Crypt/Cipher/OpenSSLAeadCipher.php]: If you pass a tag shorter than 16 bytes, openssl_decrypt() still computes a 16-byte tag internally - and it will never match the shorter one.
